internal/rbd/snapshot.go

```diff
@@ -176,3 +179,163 @@ func undoSnapshotCloning(
 
 	return err
 }
+
+// NewSnapshotByID creates a new rbdSnapshot from the rbdVolume.
+//
+// Parameters:
+// - name of the new rbd-image backing the snapshot
+// - id of the rbd-snapshot to clone
+//
+// FIXME: When resolving the Snapshot, the RbdImageName will be set to the name
+// of the parent image. This is can cause issues when not accounting for that
+// and the Snapshot is deleted; instead of deleting the snapshot image, the
+// parent image is removed.
+//
+//nolint:gocyclo,cyclop // TODO: reduce complexity.
+func (rv *rbdVolume) NewSnapshotByID(
+	ctx context.Context,
+	cr *util.Credentials,
+	name string,
+	id uint64,
+) (types.Snapshot, error) {
+	snap := rv.toSnapshot()
+	snap.RequestName = name
+
+	srcVolID, err := rv.GetID(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	snap.SourceVolumeID = srcVolID
+
+	// reserveSnap sets snap.{RbdSnapName,ReservedID,VolID}
+	err = reserveSnap(ctx, snap, rv, cr)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create a reservation in the journal for snapshot image %q: %w", snap, err)
+	}
+	defer func() {
+		if err != nil {
+			undoErr := undoSnapReservation(ctx, snap, cr)
+			if undoErr != nil {
+				log.WarningLog(ctx, "failed undoing reservation of snapshot %q: %v", name, undoErr)
+			}
+		}
+	}()
+
+	// A new snapshot image will be created, and needs to have a unique
+	// name.
+	// FIXME: the journal contains rv.RbdImageName as SourceName. When
+	// resolving the snapshot image, snap.RbdImageName will be set to the
+	// original RbdImageName/SourceName (incorrect). This is fixed-up in
+	// rbdManager.GetSnapshotByID(), this needs to be done cleaner.
+	snap.RbdImageName = snap.RbdSnapName
+
+	err = rv.Connect(cr)
+	if err != nil {
+		return nil, err
+	}
+
+	err = rv.openIoctx()
+	if err != nil {
+		return nil, err
+	}
+
+	options, err := rv.constructImageOptions(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer options.Destroy()
+
+	err = options.SetUint64(librbd.ImageOptionCloneFormat, 2)
```

Comment:
what are the image features we are using for this clone? is it the same as the existing clones we create during volumesnapshot?

Reply:
The clone will have the same options as the source image, with the clone-format set to version 2.